tests/unit/test-extension-runners/test.extension-runners.js

```diff
@@ -0,0 +1,584 @@
+          return defaultReloadStrategy(
+            {...args, ...argOverride},
+            {...options, ...optOverride});
+        },
+      };
+    }
+
+    it('configures a watcher', () => {
+      const {
+        createWatcher, reloadStrategy,
+        ...sentArgs
+      } = prepare();
+
+      reloadStrategy();
+      sinon.assert.called(createWatcher);
+      sinon.assert.calledWith(
+        createWatcher,
+        sinon.match({
+          client: sentArgs.client,
+          sourceDir: sentArgs.sourceDir,
+          artifactsDir: sentArgs.artifactsDir,
+          ignoreFiles: sentArgs.ignoreFiles,
+        })
+      );
+    });
+
+    it('configure the watcher to reload an extension by sourceDir', () => {
+      const {
+        extensionRunner, createWatcher, reloadStrategy,
+      } = prepare({
+        stubExtensionRunner: {
+          reloadExtensionBySourceDir() {},
+        },
+      });
+
+      reloadStrategy();
+
+      sinon.assert.calledOnce(createWatcher);
+      sinon.assert.calledWith(
+        createWatcher,
+        sinon.match({
+          reloadExtension: sinon.match.typeOf('function'),
+        })
+      );
+
+      const sourceDir = '/fake/sourceDir';
+      const {reloadExtension} = createWatcher.firstCall.args[0];
+      reloadExtension(sourceDir);
+
+      const {reloadExtensionBySourceDir} = extensionRunner;
+      sinon.assert.calledOnce(reloadExtensionBySourceDir);
+      sinon.assert.calledWith(
+        reloadExtensionBySourceDir,
+        sinon.match(sourceDir)
+      );
+    });
+
+    it('cleans up when the extension runner closes', () => {
+      const {
+        extensionRunner, watcher, reloadStrategy, stdin,
+      } = prepare({
+        stubExtensionRunner: {
+          registerCleanup() {},
+        },
+      });
+
+      sinon.spy(stdin, 'pause');
+
+      reloadStrategy();
+
+      sinon.assert.called(extensionRunner.registerCleanup);
+      sinon.assert.calledOnce(extensionRunner.registerCleanup);
+      sinon.assert.calledWith(
+        extensionRunner.registerCleanup,
+        sinon.match.typeOf('function')
+      );
+
+      const registeredCb = extensionRunner.registerCleanup.firstCall.args[0];
+      registeredCb();
+
+      sinon.assert.called(watcher.close);
+      sinon.assert.called(stdin.pause);
+    });
+
+    it('can reload when user presses R in shell console', async () => {
+      const {extensionRunner, reloadStrategy} = prepare();
+
+      const fakeStdin = new tty.ReadStream();
+      sinon.spy(fakeStdin, 'setRawMode');
+      sinon.spy(extensionRunner, 'reloadAllExtensions');
+
+      try {
+        await reloadStrategy({}, {stdin: fakeStdin});
+        fakeStdin.emit('keypress', 'r', {name: 'r', ctrl: false});
+
+        // Wait for one tick.
+        await Promise.resolve();
+
+        sinon.assert.called(fakeStdin.setRawMode);
+        sinon.assert.called(extensionRunner.reloadAllExtensions);
+      } finally {
```

Comment:
@kumar303 now that these tests are implemented as async functions, I opted for `try...finally` as the strategy to ensure that we always exit the "keypress processing loop" (instead of the previous strategy based on the `loopExiter` and `exitLoopOnError` test helpers).